### PR TITLE
[HandshakeToFIRRTL] Add MemoryOp lowering.

### PR DESCRIPTION
**Pull request by mikeurbach**
This adds the start of a lowering from MemoryOp. This initial commit
includes support for the load ports of a memory. The appropriate data
signals are connected, and the control signals are built similarly to
the ForkOp. This is part of https://github.com/llvm/circt/issues/337.

### CODE REVIEW

**Comment by mikeurbach** [comment]
@hanchenye this is the first part, so I thought I'd share it and make a PR. I am continuing to work on the lowering for the store ports next.

**Comment by mikeurbach** [comment]
While implementing the lowering of stores, I noticed nothing is driving the load enable on the memory in this patch. I will fix that in the next patch.

**Comment by mikeurbach** [comment]
@hanchenye I've added the lowering for stores as well. Please feel free to review these as separate patches is if that is easier.

**Comment by lattner** [comment]
Code looks nice from a style perspective, nice work @mikeurbach !

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 1]
```diff
@@ -1322,6 +1323,245 @@ bool HandshakeBuilder::visitHandshake(BufferOp op) {
...
+    auto bufferName = rewriter.getStringAttr("writeValidBuffer");
+    auto writeValidBuffer = rewriter.create<RegInitOp>(
+        insertLoc, bitType, clock, reset, falseConst, bufferName);
+
+    // Create a gate for when both the store address and data are valid.
+    auto writeValid = rewriter.create<AndPrimOp>(
+        insertLoc, bitType, storeAddrValid, storeDataValid);
+
+    // Connect the write valid signal to the write valid buffer.
+    rewriter.create<ConnectOp>(insertLoc, writeValidBuffer, writeValid);
```
Comment (hanchenye) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1530]: I think this buffer should be similar to [Fig.8](http://www1.cs.columbia.edu/~sedwards/papers/edwards2019compositional.pdf) which will only be updated when the output is ready or the buffer itself is empty?

**Review by hanchenye — COMMENTED** [review]
Very nice implement, thanks!

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 2]
```diff
@@ -1322,6 +1323,245 @@ bool HandshakeBuilder::visitHandshake(BufferOp op) {
...
 }
 
+bool HandshakeBuilder::visitHandshake(MemoryOp op) {
+  // Get the memory type and element type.
+  MemRefType type = op.type();
+  Type elementType = type.getElementType();
+  if (!elementType.isIntOrFloat()) {
+    op.emitError("only memrefs of ints or floats are supported");
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1333]: That's a good point. I will restrict this to ints for now and we can have a larger discussion about support for floating/fixed/etc. types.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 1 above)
Comment (mikeurbach) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1530]: Yes, that makes sense. I will update this accordingly.

**Review by hanchenye — APPROVED** [review]
LGTM

**Comment by mikeurbach** [comment]
Thanks @hanchenye I will merge after fixing the rebase.

**Pull request merged**